### ISSUE TITLE
A general question: Why not using the multilingual_translation task directly?

### ISSUE THREAD

**Issue by ever4244** [raymondhs/fairseq-laser#5]
Hello Raymond:

Long time no see!

I have a general question: Why not using the multilingual_translation task directly?
for example, would a setting like this take a similar effect? with a few modifications to the laser_lst model?
   --task multilingual_translation --arch laser \
    --lang-pairs de-en,de-es,en-es,es-en,fr-en,fr-es \
    --share-decoders --share-decoder-input-output-embed \

I am asking this is because I also implemented a baseline of transformer structure with the setting below
 --task multilingual_translation --arch multilingual_transformer_iwslt_de_en \
    --lang-pairs de-en,de-es,en-es,es-en,fr-en,fr-es \
    --share-decoders --share-decoder-input-output-embed \

What I want to achieve is to compare the difference between LASER's LSTM structure with Transformer's attention based structure. When I doing this, I find the only major difference between the laser task and multilingual_translatio  task is that the former uses a "multi_corpus_sampled_dataset" based laser data while the latter uses a "round_robin_zip_datasets dataset.   And multilingual_translation provides a more general purposed setting. 

So would there be a performance difference between round_robin_zip and multi_corpus_sampled
method in this task? I think the sampling is uniform, therefore, in theory, they should be roughly the same?

Why do you want to implement a LASER task instead of directly using the default multilingual_translation task provided by fairseq, and setting encoder and decoder shared by all languages and using the same dictionary? Do I miss something?

Thank you very much!

**Comment by raymondhs** [comment]
Hello,

The main difference is in the training loop, which is roughly:

With RoundRobinZipDatasets:
```
for i in range(len(epoch)):
  for lang_pair in args.lang_pairs:
    batch = next_batch_for_lang_pair(lang_pair)
    loss = criterion(model_for_lang_pair(lang_pair), batch)
    loss.backward()
  optimizer.step()
```

With MultiCorpusSampledDataset:
```
for i in range(len(epoch)):
  lang_pair = sample_one_lang_pair(args.lang_pairs)
  batch = next_batch_for_lang_pair(lang_pair)
  loss = criterion(model_for_lang_pair(lang_pair), batch)
  loss.backward()
  optimizer.step()
```

So effectively, you have a smaller total number of model updates with bigger batch size (multiple batches from all lang pairs) with RoundRobinZipDatasets, vs. a bigger number of updates with smaller batch size (one batch from just 1 random lang pair) with MultiCorpusSampledDataset.

My initial implementation (before January 2020) was closer to fairseq's multilingual_translation, except you also need to include language ID embeddings. I changed to MultiCorpusSampledDataset after reading from the [LASER's blog post](https://engineering.fb.com/ai-research/laser-multilingual-sentence-embeddings/), which stated that:

> We trained our system on 223 million sentences of public parallel data, aligned with either English or Spanish. For each mini-batch, we randomly chose an input language and trained the system to translate the sentences into English or Spanish.

I managed to get a closer result to LASER's paper using this way, but I would think that both ways are viable, it's kind of empirical. :)

**Comment by ever4244** [comment]
Thank you for the insight
Does the performance difference a significant one?

multilingual_translation task is attractive for me because I can choose whether to use a shared encoder or decoder easily, this made a more flexible model structure viable. And I can do a batch of experiments on the different model structures using the same task framework.

Yes, I notice that the two methods are mainly different at how to do the mini-batch and update. 
So would it a viable option for me to use your laser_dataset, in combination with the multilingual_translation task?

I notice the three major differences compared with the standard multilingual_translation setting that I need to add to my current code, one is the laser_dataset (MultiCorpusSampledDataset), the other is the language id on the decoder input, and finally a sentence embedding concatenation to the decoder input.
 
 Is there anything else that I should be aware of?
                                                                                                            Regards!

**Comment by ever4244** [comment]
BTW:
I also find in your code the decoder forward() function takes in a decoder language as input.
And it then embeds the language.
However, it appears that l is not being used? Is that normal?
Does it seem that language ID embedding is not necessary?

`   # embed language
        lang_tensor = torch.LongTensor(
            [self.lang_dictionary[lang]] * bsz
        ).to(device=prev_output_tokens.device)
        l = self.embed_langs(lang_tensor)`

`class LaserDecoder(FairseqIncrementalDecoder):

    def __init__(
        self, dictionary, lang_dictionary, embed_dim=512, hidden_size=512, out_embed_dim=512,
        num_layers=1, dropout_in=0.1, dropout_out=0.1, attention=True,
        encoder_output_units=512, pretrained_embed=None,
        share_input_output_embed=False, adaptive_softmax_cutoff=None,
        lang_embedding_size=32
    ):
        super().__init__(dictionary)
        self.dropout_in = dropout_in
        self.dropout_out = dropout_out
        self.hidden_size = hidden_size
        self.share_input_output_embed = share_input_output_embed
        self.lang_embedding_size = lang_embedding_size
        self.lang_dictionary = lang_dictionary
        self.embed_langs = nn.Embedding(len(lang_dictionary), lang_embedding_size)
        self.need_attn = False

        self.adaptive_softmax = None
        num_embeddings = len(dictionary)
        padding_idx = dictionary.pad()
        self.embed_tokens = Embedding(num_embeddings, embed_dim, padding_idx)

        self.encoder_output_units = encoder_output_units
        self.encoder_hidden_proj = Linear(encoder_output_units, hidden_size)
        self.encoder_cell_proj = Linear(encoder_output_units, hidden_size)
        
        input_size = hidden_size + embed_dim + lang_embedding_size + encoder_output_units
        self.layers = nn.ModuleList([
            LSTMCell(
                input_size=input_size if layer == 0 else hidden_size,
                hidden_size=hidden_size,
            )
            for layer in range(num_layers)
        ])
        
        self.attention = None
        if hidden_size != out_embed_dim:
            self.additional_fc = Linear(hidden_size, out_embed_dim)
        if not self.share_input_output_embed:
            self.fc_out = Linear(out_embed_dim, num_embeddings, dropout=dropout_out)

    def forward(self, prev_output_tokens, encoder_out, lang, incremental_state=None):
        x, attn_scores = self.extract_features(
            prev_output_tokens, encoder_out, lang, incremental_state
        )
        return self.output_layer(x), attn_scores

    def extract_features(
        self, prev_output_tokens, encoder_out, lang, incremental_state=None
    ):
        """
        Similar to *forward* but only return features.
        """
        encoder_sentemb = encoder_out['sentemb']
        encoder_padding_mask = encoder_out['encoder_padding_mask']
        encoder_out = encoder_out['encoder_out']

        if incremental_state is not None:
            prev_output_tokens = prev_output_tokens[:, -1:]
        bsz, seqlen = prev_output_tokens.size()

        # get outputs from encoder
        encoder_outs, encoder_hiddens, encoder_cells = encoder_out[:3]
        srclen = encoder_outs.size(0)

        # embed tokens
        x = self.embed_tokens(prev_output_tokens)
        x = F.dropout(x, p=self.dropout_in, training=self.training)

        # B x T x C -> T x B x C
        x = x.transpose(0, 1)

        # embed language
        lang_tensor = torch.LongTensor(
            [self.lang_dictionary[lang]] * bsz
        ).to(device=prev_output_tokens.device)
        l = self.embed_langs(lang_tensor)

        # initialize previous states (or get from cache during incremental generation)
        cached_state = utils.get_incremental_state(self, incremental_state, 'cached_state')
        if cached_state is not None:
            prev_hiddens, prev_cells, input_feed = cached_state
        else:
            num_layers = len(self.layers)
            prev_hiddens = [encoder_sentemb for i in range(num_layers)]
            prev_cells = [encoder_sentemb for i in range(num_layers)]
            prev_hiddens = [self.encoder_hidden_proj(x) for x in prev_hiddens]
            prev_cells = [self.encoder_cell_proj(x) for x in prev_cells]
            input_feed = x.new_zeros(bsz, self.hidden_size)

        outs = []
        for j in range(seqlen):
            # input feeding: concatenate context vector from previous time step
            input = torch.cat((x[j, :, :], encoder_sentemb, input_feed, l), dim=1)

            for i, rnn in enumerate(self.layers):
                # recurrent cell
                hidden, cell = rnn(input, (prev_hiddens[i], prev_cells[i]))

                # hidden state becomes the input to the next layer
                input = F.dropout(hidden, p=self.dropout_out, training=self.training)

                # save state for next time step
                prev_hiddens[i] = hidden
                prev_cells[i] = cell

            out = hidden
            out = F.dropout(out, p=self.dropout_out, training=self.training)

            # input feeding
            input_feed = out

            # save final output
            outs.append(out)

        # cache previous states (no-op except during incremental generation)
        utils.set_incremental_state(
            self, incremental_state, 'cached_state',
            (prev_hiddens, prev_cells, input_feed),
        )

        # collect outputs across time steps
        x = torch.cat(outs, dim=0).view(seqlen, bsz, self.hidden_size)

        # T x B x C -> B x T x C
        x = x.transpose(1, 0)

        if hasattr(self, 'additional_fc') and self.adaptive_softmax is None:
            x = self.additional_fc(x)
            x = F.dropout(x, p=self.dropout_out, training=self.training)

        return x, None`

**Comment by raymondhs** [comment]
I don't recall doing the BUCC experiments using the multilingual_translation task so I am not too sure about the performance difference. My implementation for LaserDataset is based on MultiCorpusSampledDataset, so it may not be suitable to use directly with Fairseq's multilingual_task (it needs RoundRobinZipDatasets). I guess you can check multilingual_translation.py in Fairseq, and find some ways to include the language ID information when calling your model.

**Comment by ever4244** [comment]
> I don't recall doing the BUCC experiments using the multilingual_translation task so I am not too sure about the performance difference. My implementation for LaserDataset is based on MultiCorpusSampledDataset, so it may not be suitable to use directly with Fairseq's multilingual_task (it needs RoundRobinZipDatasets). I guess you can check multilingual_translation.py in Fairseq, and find some ways to include the language ID information when calling your model.


What I did is to tie language id with sentemb from the encoder, and shut down the real encoder output to the decoder (I used a transformer structure). So instead of attention on the encoder output, the decoder attention only attention on the sentemb and language id.

I once tie the language id with x (the decoder side input), but then I thought it would be faster and cheaper to just concatenate sentemb with language embedding.  

Because I am using a new structure (transformer) I am testing different ways to adapt laser into that. 
Currently, I am considering whether I should use "sharing encoder and decoder input embedding" Have you ever used --share-decoder-input-output-embed?

Thank you for the help!
Will share to you my finding if you are interested, once all my experiments are done.
